### ISSUE TITLE
Transcription difference between icefall and sherpa-onnx

### ISSUE THREAD

**Issue by pharma-vie** [k2-fsa/icefall#2005]
I have transcription problems after exporting the model which become catastrophic either an insertion or deletion or substitution
I trained a Zipformer streaming model with the [everyayah ](https://huggingface.co/datasets/tarteel-ai/everyayah) dataset using the following command:
```
training_script_path = ICEFALL_DIR / "egs/librispeech/ASR/pruned_transducer_stateless7_streaming/train.py"

cmd_args = [
    "python3", str(training_script_path),
    "--world-size=1",
    f"--num-epochs={NUM_EPOCHS}",
    f"--exp-dir={exp_dir_current_stage}",
    f"--max-duration={MAX_DURATION}",
    "--use-fp16=1",
    f"--base-lr={BASE_LR}",
    f"--bpe-model={DRIVE_BPE_DIR}/everyayah_bpe_500.model",
    f"--manifest-dir={UNIFIED_MANIFEST_DIR}",
    f"--num-workers={NUM_WORKERS}",
    "--enable-musan=False",
    "--full-libri=0",
    "--save-every-n=5000",
    "--short-chunk-size=50",
    "--num-left-chunks=4",
    "--decode-chunk-len=32",
]
```

.
.

**1- The result of decode.py:**

 ```
CHUNK_SIZE_KEY = 32
  decoding_script_path = "./pruned_transducer_stateless7_streaming/decode.py"
    cmd_args = [
        "python3", decoding_script_path,
        f"--iter={EPOCH_TO_TEST}",
        f"--avg={AVG}",
        f"--use-averaged-model={USE_AVERAGED_MODEL}",
        f"--exp-dir={EXP_DIR}",
        f"--bpe-model={BPE_MODEL_PATH}",
        f"--decoding-method={DECODING_METHOD}",
        f"--max-duration={MAX_DURATION}",
        f"--decode-chunk-len={CHUNK_SIZE_KEY}",
        f"--manifest-dir={PREPARED_MANIFEST_DIR}",
    ]
```

**with greedy_search**	2.02	best for test-clean
**greedy_search**	2.02	best for test-other

**%**WER** = 2.02**
Errors: 15 insertions, 163 deletions, 69 substitutions, over 12231 reference words (11999 correct)

**with modified_beam_search**
**beam_size_4**	1.71	best for test-clean
**beam_size_4**	1.71	best for test-other

**%WER = 1.71**
Errors: 17 insertions, 149 deletions, 43 substitutions, over 12231 reference words (12039 correct)

.
.

**2- The result of streaming_decode.py:**

```
    decoding_script_path = "./pruned_transducer_stateless7_streaming/decode.py"
    cmd_args = [
        "python3", decoding_script_path,
        f"--iter={ITER_TO_TEST}",
        f"--avg={AVG}",
        f"--use-averaged-model={USE_AVERAGED_MODEL}",
        f"--exp-dir={EXP_DIR}",
        f"--bpe-model={BPE_MODEL_PATH}",
        f"--decoding-method={DECODING_METHOD}",
        f"--max-duration={MAX_DURATION}",
        f"--decode-chunk-len={CHUNK_SIZE_KEY}",
        f"--manifest-dir={PREPARED_MANIFEST_DIR}",
    ]
```

**greedy_search**	1.99	best for test-clean
**greedy_search**	1.99	best for test-other

**%WER = 1.99**
Errors: 15 insertions, 159 deletions, 70 substitutions, over 12231 reference words (12002 correct)

**with modified_beam_search**
**num_active_paths_4**	1.69	best for test-clean
**num_active_paths_4**	1.69	best for test-other

%WER = 1.69
Errors: 18 insertions, 143 deletions, 46 substitutions, over 12231 reference words (12042 correct)

.
.

**3- I exported the pretrained model with sherpa-onnx:**
with commande

  ```
./pruned_transducer_stateless7_streaming/export-onnx.py \\
      --iter "{iter_to_export}" \\
      --decode-chunk-len=32 \\
      --avg 1 \\
      --use-averaged-model false \\
      --exp-dir "{exp_dir}" \\
      --tokens "{tokens_path}"
    """
```

.

**here it is a deletion**
```
python /content/sherpa-onnx/python-api-examples/online-decode-files.py \
      --tokens="/content/drive/MyDrive/everyayah_R_zipformer/bpe_model/tokens.txt" \
      --encoder="/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2/encoder-iter-125000-avg-1.onnx" \
      --decoder="/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2/decoder-iter-125000-avg-1.onnx" \
      --joiner="/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2/joiner-iter-125000-avg-1.onnx" \
      --decoding-method=modified_beam_search \
      --num-threads=3 \
      "/content/drive/MyDrive/everyayah_R_zipformer/wav/test-00000-of-00013-18d88ded65aa7b53_1078.wav"
    
--------------------------------------------------------------------------------
Started!
Done!
/content/drive/MyDrive/everyayah_R_zipformer/wav/test-00000-of-00013-18d88ded65aa7b53_1078.wav
ةً لِلَّذِينَ كَفَرُوا وَاغْفِرْ لَنَا رَبَّنَا إِنَّكَ أَنْتَ الْعَزِيزُ الْحَكِيمُ
----------
num_threads: 3
decoding_method: modified_beam_search
Wave duration: 32.287 s
Elapsed time: 19.345 s
Real time factor (RTF): 19.345/32.287 = 0.599
```

**recogs-test-clean-iter-125000-avg-1-streaming-chunk-size-32.txt**
**test-00000-of-00013-18d88ded65aa7b53_1078-1078**:	**ref**=['رَبَّنَا', 'لَا', 'تَجْعَلْنَا', 'فِتْنَةً', 'لِلَّذِينَ', 'كَفَرُوا', 'وَاغْفِرْ', 'لَنَا', 'رَبَّنَا', 'إِنَّكَ', 'أَنْتَ', 'الْعَزِيزُ', 'الْحَكِيمُ']
**test-00000-of-00013-18d88ded65aa7b53_1078-1078**:	**hyp**=['رَبَّنَا', 'لَا', 'تَجْعَلْنَا', 'فِتْنَةً', 'لِلَّذِينَ', 'كَفَرُوا', 'وَاغْفِرْ', 'لَنَا', 'رَبَّنَا', 'إِنَّكَ', 'أَنْتَ', 'الْعَزِيزُ', 'الْحَكِيمُ']

.
.

**here it is a substitution**
python /content/sherpa-onnx/python-api-examples/online-decode-files.py \
      --tokens="/content/drive/MyDrive/everyayah_R_zipformer/bpe_model/tokens.txt" \
      --encoder="/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2/encoder-iter-125000-avg-1.onnx" \
      --decoder="/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2/decoder-iter-125000-avg-1.onnx" \
      --joiner="/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2/joiner-iter-125000-avg-1.onnx" \
      --decoding-method=modified_beam_search \
      --num-threads=3 \
      "/content/drive/MyDrive/everyayah_R_zipformer/wav/test-00000-of-00013-18d88ded65aa7b53_1141.wav"
    
--------------------------------------------------------------------------------
Started!
Done!
/content/drive/MyDrive/everyayah_R_zipformer/wav/test-00000-of-00013-18d88ded65aa7b53_1141.wav
**وَإِنَّ الْفُ الشارَ لَ سَبِيِي جَحِيمٍ**
----------
num_threads: 3
decoding_method: modified_beam_search
Wave duration: 12.539 s
Elapsed time: 7.241 s
Real time factor (RTF): 7.241/12.539 = 0.578

**recogs-test-clean-iter-125000-avg-1-streaming-chunk-size-32.txt**
test-00000-of-00013-18d88ded65aa7b53_1141-1141:	ref=['وَإِنَّ', 'الْفُجَّارَ', 'لَفِي', 'جَحِيمٍ']
test-00000-of-00013-18d88ded65aa7b53_1141-1141:	hyp=['وَإِنَّ', 'الْفُجَّارَ', 'لَفِي', 'جَحِيمٍ']

**Comment by csukuangfj** [comment]
Please share the value of
- iter to decode and avg
- iter to test and avg
- iter to export and avg

**Comment by pharma-vie** [comment]
At first, the value of iter was 125000 with avg 1.
I redid the decoding with decode.py with avg 5, and the WER improved.
greedy_search 1.13
beam_size_4 1.22
but the problem persists.

- `python3 ./pruned_transducer_stateless7_streaming/decode.py
-  --iter=125000
-  --avg=5
-  --use-averaged-model=True
-  --exp-dir=/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2
-  --bpe-model=/content/drive/MyDrive/everyayah_R_zipformer/bpe_model/everyayah_bpe_500.model 
- --decoding-method=greedy_search
-  --max-duration=100
-  --decode-chunk-len=32
-  --manifest-dir=/content/manifests_for_decoding`


.
.


 ./pruned_transducer_stateless7_streaming/export-onnx.py \
      --iter "125000" \
      --decode-chunk-len=32 \
      --avg 5 \
      --use-averaged-model true \
      --exp-dir "/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2" \
      --tokens "/content/drive/MyDrive/everyayah_R_zipformer/bpe_model/tokens.txt"
    


.
.

   python /content/sherpa-onnx/python-api-examples/online-decode-files.py \
     --tokens="/content/drive/MyDrive/everyayah_R_zipformer/bpe_model/tokens.txt" \
     --encoder="/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2/encoder-iter-125000-avg-5-with-averaged-model.onnx" \
      --decoder="/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2/decoder-iter-125000-avg-5-with-averaged-model.onnx" \
     --joiner="/content/drive/MyDrive/everyayah_R_zipformer/icefall_exp/zipformer_full_data_A100_v2/joiner-iter-125000-avg-5-with-averaged-model.onnx" \
     --decoding-method=greedy_search  \
      --num-threads=3 \
     "/content/drive/MyDrive/everyayah_R_zipformer/wav/test-00000-of-00013-18d88ded65aa7b53_1141.wav"

.
.

**Comment by csukuangfj** [comment]
> but the problem persists.

How many audio files have you tested with sherpa-onnx and do all tested files have inconsistent results with icefall? 

Can you also compute the WER with sherpa-onnx?

**Comment by joazoa** [comment]
@pharma-vie It's a bit difficult for understand your issue, can you give the results a bit more structured ?

E.g.
AVG_1, chunksize 32, test set WER = X LER = Y, .pt, 
AVG_1, chunksize 32, test set WER = X LER = Y, .onnx
AVG_1, chunksize 32, test set WER = X LER = Y, .int8.onnx

And then some typical examples ?

We have seen some issues with the .onnx being less accurate, but its mostly the int8 or large chunksizes we have issues with. The difference we see is mostly deletions, adding a blankscore penalty may help.
